corehq/apps/case_importer/do_import.py

```diff
@@ -250,6 +259,17 @@ def _populate_updated_fields(self, row):
             elif update_value is not None:
                 update_value = _convert_field_value(update_value)
 
+            if update_field_name in self.validated_fields:
+                case_property = self.validated_fields[update_field_name]
+                if not case_property.valid_value(update_value):
+                    # note: probably want to accumulate full list of invalid
+                    # columns for this row and raise an exception with that
+                    # full set of info rather than bailing out on first one
+                    # as we're currently doing here.
+                    if case_property.data_type == 'date':
```

Comment:
It'd be good to have the data type conditionals in only one place, so that someone adding a new type of validation has to update as few places as possible. This also seems like it could get gnarly if there were multiple kinds of errors for a single data type.

Could this logic go into `valid_value`, which would then probably raise an exception on failure and return None/True on success?

Reply:
Yeah, my initial intent was to have all the validation details over in the data dictionary side, that got derailed a bit when discovering the existing setup for collecting/reporting errors on import relied on customized exceptions defined in the case importer app. I suppose it would be OK for the data dictionary code to raise the exceptions defined in the case importer app, though it does tie the provider (data dictionary) of the validation more to the "client" using it (case importer) than I would have hoped at the start.

Reply:
The tighter coupling is a good point. As mentioned the last time we met, I think there are case importer exceptions that may make more sense as part of the data dictionary, but that are in the case importer just because that code is much older. I think the benefit of having all the validation in one place is worth the tighter coupling, and it should be pretty easy to move the exceptions from one app to another in the future.

Reply:
Ok I went this route of having the data dictionary raise the exception defined in case importer. One wrinkle is that exception wants the field name, which is rather superfluous to send into the validity checker, so it's added by the calling code after the exception is raised. There is other code in do_import that does similar for title and message after creating a generic `CaseRowError` ... not quite the same pattern but similar. Alternatively the column name could be passed to the validity checker solely for the purpose of creating the exception with the right info to begin with. Either works, I'm not sure which would be better.